README.md

```diff
@@ -1,17 +1,170 @@
+
+``` ipldsch
+type Principal union {
+  -- Specification defines principals in terms of did:key's (that are multicodec
+  -- identifier for the public key type followed by the raw bytes of the key)
+  -- we represent those as raw public key bytes prefixed with public key multiformat
+  -- code.
+  | Ed25519    "0xed"
+  | RSA        "0x1205"
+  | P256       "0x1200"
+  | P384       "0x1201"
+  | P521       "0x1202"
+  
+  -- To accomodate additional DID methods we represent those as UTF8 encoding
+  -- of the DID omitting `did:` prefix itself. E.g. `did:dns:ucan.xyz` can be
+  -- represented as [0xd1d, ...new TextEncoder().encode('dns:ucan.xyz')] bytes.
+  | DID "0xd1d"
+} representation bytesprefix
+```
+
+## 2.2 Capability
+
+``` ipldsch
+type Capability struct {
+  with Resource
+  can Ability
+  nb optional NonNormativeFields
+}
+```
+
+### 2.2.1 Resource
+
+The resource pointer MUST be formatted as a [URI](https://www.rfc-editor.org/rfc/rfc3986).
+
+``` ipldsch
+type Resource = String
+```
+
+### 2.2.2 Ability
+
+Abilities MUST be lowercase, and MUST be namespaced with `/` delimeters. Abilities SHOULD have at least one path segment.
+
+``` ipldsch
+type Ability = String
+```
+
+### 2.2.3 `nb` Non-Normative Fields
+
+The `nb` capability field is OPTIONAL. When present, it MUST contain any additional domain specific details and/or restrictions of the capability. 
+
+``` ipldsch
+type NonNormativeFields = { String: Any }
+```
+
+## 2.3 Fact
+
+Facts MUST be structured as maps with string keys.
+
+``` ipldsch
+type Fact { String: Any }
+```
+
+## 2.4 Proof
+
+As of UCAN 0.9, all proofs MUST be given as links (CIDs). Note that UCANs MAY be inlined via the [identity multihash (`0x00`)](https://github.com/multiformats/multicodec/blob/master/table.csv#L2) CID.
+
+``` ipldsch
+&UCAN
+```
+
+## 2.5 Signature
+
+The signature MUST be computed by first encoding it as a [canonical JWT](#3-jwt-canonicalization), and then signed with the issuer's private key. Signatures MUST be encoded as the _varsig_ multiformat representation `<varint sig_alg_code><vairint sig_size><bytes sig_output>`. _Varsig_ multiformat codes SHOULD be [registered multicodec codes](https://github.com/multiformats/multicodec). Unregistered signature types MAY be added via the `NonStandardSignature` prefix.
+
+``` ipldsch
+type Signature union {
+  -- Algorithms here are expected to be valid "varsig" multiformat codes.
+  | NonStandard "0xd000"
+  | EdDSA       "0xd0ed"
+  | RS256       "0xd125"
+  | ES256       "0xd120"
+  | ES256K      "0xd0e7" -- secp256k1
+  | BLS12381    "0xd0ea"
+  | EIP191      "0xd191"
+} representation bytesprefix
```

Comment:
If this is too cute, I can go back to numbering them `0xd001`, `0xd002`, etc.

## Punning

* Ed25519
  * Public key prefix: `0xed`
  * Signature prefix: `0xd0ed`
* Not enough characters for RS256
  * Public key prefix: `0x1205`
  * Signature prefix: `0xd125`
* EIP-191 (alt signature for secp256k1)
  * Public key: `0xe7`
  * Signature: `0xd191`